src/librustc/ty/inhabitedness/mod.rs

```diff
@@ -134,9 +141,16 @@ impl<'a, 'gcx, 'tcx> VariantDef {
             AdtKind::Enum => true,
             AdtKind::Struct => false,
         };
-        DefIdForest::union(tcx, self.fields.iter().map(|f| {
-            f.uninhabited_from(tcx, substs, is_enum)
-        }))
+        let is_declared_nonexhaustive =
+            self.is_field_list_non_exhaustive() && !self.def_id.is_local();
+        // Non-exhaustive variants from other crates are always considered inhabited.
+        if is_enum && is_declared_nonexhaustive {
```

Comment:
The `is_enum` condition is incorrect, structs can also break if a new uninhabited field is added (https://github.com/rust-lang/rust/pull/60529#issuecomment-489429583).